### PR TITLE
Create assets registry

### LINKED ISSUE #6353: Asset Management - Add Assests Registry page
## WARNING:  Issues https://github.com/IMA-WorldHealth/bhima/issues/6348, and https://github.com/IMA-WorldHealth/bhima/issues/6349, and https://github.com/IMA-WorldHealth/bhima/issues/6352 should be done before this issue!

- Add “Assets Registry” page under the "Assets Management" top-level menu item
  - Brings up the Stock Lot registry with flags to show asset-related fields by default.  We might change the Title to “Assets” vs “Lots” in this case.

### PR DESCRIPTION
**Pull request by jmcameron**
**This PR**
- Creates an "assets" registry (under the top-level "Assets Management" menu
- Merges in the old stock assignment functionality into this registry
- Adds one Asset inventory group:  Véhicle de transport (Code:  E.VEH)
- Adds one Asset inventory article: Code: MOT.HCRF250RX (Honda CRF250RX motorcycle)
- Updates tests for the new inventory group & item
- Removes old "Stock Assignments" page under the "Stock" menu (since all of its functionality has been added to the Assets Registry)
- Fixes missing translation items (some of these were not related to Assets but fixed old missing items)
- Fixed the menu bar item order for consistency in a couple of pages and colorized "Add" buttons

Closes https://github.com/IMA-WorldHealth/bhima/issues/6353

**TESTING**
- use bhima_test
- Verify the new Inventory group and article type in the Inventory Configuration and Registry pages
- Use Stock Entry to enter a new motorcycle (see above)
- Go to the "Asset Management > Asset Registry" page
   - Verify the new asset (lot) is present
   - Try editing it via the action menu
   - Try adding an assignment via the big [Add Assignment] button on the menu bar
   - Verify the new assignment
   - Use "Remove Assignment" in the action menu to remove the assignment
   - Use the "Add Assignment" link in the action menu to add a new assignment
   - Use the "Assignment History" link in the action menu to see the past assignments
   - Check the other action items (barcode, etc)

- Double-check all new French translations and suggest updates as needed

- Try using a production database and verify that the "Stock > Stock Assignment" page is no longer available.

### CODE REVIEW

File: client/src/i18n/fr/settings.json [hunk 1]
```diff
@@ -3,6 +3,8 @@
     "BASE_INDEX_GROWTH_RATE_HELP_TEXT" : "Le système de paie avec Indice prend généralement en compte l'indice de base de départ à l'embauche, la date d'embauche et leur niveau d'éducation pour projeter leur indice de « base » (à un pourcentage d'augmentation chaque année)..",
     "BUG_LINK": "Signaler un Problème",
     "BUG_REPORT": "Merci d'avoir déposé une erreur! Veuillez lire le message suivant et remplacer le texte entouré de parenthèses ([[]]) avec votre rapport.\r\nD'où vient l'erreur?\r\n\t\t[[ ENTER TEXT HERE ]]\r\nQuelle tâche essayiez-vous d'accomplir??\r\n\t\t[[ ENTER TEXT HERE ]]\r\nQuel résultat vous attendiez-vous?\r\n\t\t[[ ENTER TEXT HERE ]]\r\nToute autre information ou commentaire pertinent?\r\n\t\t[[ ENTER TEXT HERE ]]\r\nPuissions-nous vous contacter pour plus d'informations?\r\n\t\t[x] YES [ ] NO",
+    "DEFAULT_COST_CENTER_FOR_STOCK_LOSS": "Centre de coût des pertes de stock",
```
Comment (jniles) [inline client/src/i18n/fr/settings.json:6]: Is this a mistake?  I can't figure out why you have added this and I'm pretty sure it is found somewhere else.

File: client/src/js/services/receipts/ReceiptModal.js [hunk 2]
```diff
@@ -364,13 +364,13 @@ function ReceiptModal(Modal, Receipts, Invoice, Cash, Voucher) {
   }
 
   /**
-   * @method stockAssignReceipt
+   * @method stockAssignmentReceipt
    * @param {string} uuid
    * @param {boolean} notifyCreated
    */
-  function stockAssignReceipt(uuid, notifyCreated) {
+  function stockAssignmentReceipt(uuid, notifyCreated) {
```
Comment (jniles) [inline client/src/js/services/receipts/ReceiptModal.js:371]: Good renaming 👍

File: client/src/modules/assets/assets-registry.js [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+AssetsRegistryController.$inject = [
+  'StockService', 'NotifyService', 'uiGridConstants', 'StockModalService', 'ModalService',
+  'LanguageService', 'GridGroupingService', 'GridStateService', 'GridColumnService',
+  '$state', '$httpParamSerializer', 'BarcodeService', 'AssetsRegistryService', 'bhConstants',
+  'ReceiptModal',
+];
+
+/**
+ * Stock lots Controller
```
Comment (jniles) [inline client/src/modules/assets/assets-registry.js]: Could you update these comments?

File: client/src/modules/assets/assets-registry.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+function AssetsRegistryController(
+  Stock, Notify, uiGridConstants, StockModal, Modal,
+  Languages, Grouping, GridState, Columns,
+  $state, $httpParamSerializer, Barcode, AssetsRegistry, bhConstants,
+  Receipts,
+) {
+  const vm = this;
+  const cacheKey = 'assets-grid';
+  const stockLotFilters = Stock.filter.lot;
+  const stockAssignFilters = Stock.filter.stockAssign;
```
Comment (jniles) [inline client/src/modules/assets/assets-registry.js]: Can we go all in on the renaming and make this `stockAssignmentFilters`?

File: client/src/modules/assets/modals/assign.modal.html [hunk 5]
```diff
@@ -19,19 +19,25 @@
...
         on-select="$ctrl.onSelectInventory($item)"
         ng-disabled="!$ctrl.availableInventories.length">
+
+        <!--
         <ui-select-match placeholder="{{ 'FORM.SELECT.INVENTORY' | translate }}">
           <span>{{$select.selected.hrLabel}}</span>
         </ui-select-match>
+        -->
+        <ui-select-match placeholder="{{ 'FORM.SELECT.INVENTORY' | translate }}">
+          <span>{{$ctrl.selectedInventory.hrLabel || $select.selected.hrLabel}}</span>
```
Comment (jniles) [inline client/src/modules/assets/modals/assign.modal.html:33]: What was the rational behind this change?

File: client/src/modules/assets/modals/search.modal.html [hunk 6]
```diff
@@ -1,20 +1,21 @@
 <form
   name="ModalForm"
   ng-submit="$ctrl.submit(ModalForm)"
-  data-modal="stock-assign-search"
+  data-modal="stock-lots-search"
```
Comment (jniles) [inline client/src/modules/assets/modals/search.modal.html:4]: Maybe we'll want to rename this something about `assets`?

File: client/src/modules/assets/modals/search.modal.html [hunk 7]
```diff
@@ -23,40 +24,54 @@
...
           </div>
 
-          <!-- entity -->
-          <bh-entity-select
-            entity-uuid="$ctrl.searchQueries.entity_uuid"
-            on-select-callback="$ctrl.onSelectEntity(entity)"
-            required="true">
-            <bh-clear on-clear="$ctrl.clear('entity')"></bh-clear>
-          </bh-entity-select>
+          <!-- expiration date -->
```
Comment (jniles) [inline client/src/modules/assets/modals/search.modal.html]: It looks like this is `entry_date` not expiration date.

File: client/src/modules/assets/modals/search.modal.js [hunk 8]
```diff
@@ -44,18 +39,31 @@ function SearchStockAssignModalController(data, util, Store, Instance, Periods,
...
+    displayValues.group_uuid = group.name;
   };
 
-  // assign already defined custom filters to searchQueries object
-  vm.searchQueries = util.maskObjectFromKeys(data, searchQueryOptions);
+  // Save the entry dates
+  vm.onEntryDate = (dateFrom, dateTo) => {
+    vm.searchQueries.entry_date_from = dateFrom;
+    displayValues.entry_date_from = dateFrom;
+    // if (angular.isDefined(dateFrom)) {
```
Comment (jniles) [inline client/src/modules/assets/modals/search.modal.js]: If these comments aren't necessary, go ahead and remove them.

File: client/src/modules/assets/templates/action.cell.html [hunk 9]
```diff
@@ -0,0 +1,77 @@
+<div class="ui-grid-cell-contents text-right" uib-dropdown dropdown-append-to-body uib-dropdown-toggle ng-if="!row.groupHeader">
+  <a href>
+    <span data-method="action" translate>FORM.BUTTONS.ACTIONS</span>
+    <span class="caret"></span>
+  </a>
+
+  <ul data-action="{{ rowRenderIndex}}"  class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>
```
Comment (jniles) [inline client/src/modules/assets/templates/action.cell.html]: You probably copied-pasted this from somewhere.  We don't use the `{{ rowRenderIndex }}` anymore.  Please remove it.

File: server/controllers/stock/core.js [hunk 10]
```diff
@@ -259,6 +260,22 @@ async function getLotsDepot(depotUuid, params, finalClause) {
     emptyLotToken = 'HAVING quantity = 0';
   }
 
+  // Add extra fields for assignments if the query is for assets
+  let assignmentFields = '';
+  let assignmentJoin = '';
+  if ('is_asset' in params && params.is_asset === '1') {
```
Comment (jniles) [inline server/controllers/stock/core.js:271]: I am almost a fan of adding a separate route for this.  Given how much performance is hampered by these routes. 

What do you think?  Is it worth adding a separate `stock/assets` route for this?

File: server/models/migrations/next/migrate.sql [hunk 11]
```diff
@@ -63,9 +63,16 @@ CALL add_column_if_missing('inventory_type', 'is_predefined', 'TINYINT(1) NOT NU
...
+ * @date: 2022-02-10
+ */
 CALL add_column_if_missing('lot', 'serial_number', 'VARCHAR(40) NULL');
+
+/*
+ * Issue: Asset Management - Remove old Stock Assignment page
+ * @author: jmcameron
+ * @date: 2022-03-01
+ */
+DELETE FROM unit WHERE id = 225;
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql:91]: I think you'll need to clean up permissions first.

Try deleting from `role_unit` first, then this one.

Comment (jniles) [reply server/models/migrations/next/migrate.sql:91]: You'll also need to add the new unit into this migration script.

**Review by jniles — COMMENTED** [review]
Overall, this looks really good.  I like the renaming and the cleanup work done.

There are a few bits that need addressing - migration file, some clean up comments.  I want to do a more thorough test when you've addressed those, so ping me when they are done.

File: client/src/i18n/fr/settings.json (hunk 1 above)
Comment (jmcameron) [reply client/src/i18n/fr/settings.json:6]: I ran the script to check which translation items were missing and I fixed them.  Some we not related to this PR but I fixed them anyway.    I double-checked this one and it is correct and it is the only place it was defined.  The problem before was that it was defined for English, but not French.

File: client/src/modules/assets/modals/assign.modal.html (hunk 5 above)
Comment (jmcameron) [reply client/src/modules/assets/modals/assign.modal.html:33]: If the modal is invoked with no assset selected, the item $ctrl.selectedInventory is null so you get the old behavior.  But if it is activated via the action menu, I needed a way to display the selected item.  Setting the asset inventory_uuid was not enough.   So in the controller, I set vm.selectedInventory to the asset invoking the action menu link and then the modal displays it properly.

File: server/controllers/stock/core.js (hunk 10 above)
Comment (jmcameron) [reply server/controllers/stock/core.js:271]: I'm okay with that, but I'd prefer to do it in a separate issue/PR.

File: server/models/migrations/next/migrate.sql (hunk 11 above)
Comment (jmcameron) [reply server/models/migrations/next/migrate.sql:91]: Thanks!

File: client/src/modules/assets/assets-registry.js (hunk 3 above)
Comment (jmcameron) [reply client/src/modules/assets/assets-registry.js]: Done

File: client/src/modules/assets/assets-registry.js (hunk 4 above)
Comment (jmcameron) [reply client/src/modules/assets/assets-registry.js]: Done

File: client/src/modules/assets/modals/search.modal.html (hunk 7 above)
Comment (jmcameron) [reply client/src/modules/assets/modals/search.modal.html]: Fixed

File: client/src/modules/assets/modals/search.modal.js (hunk 8 above)
Comment (jmcameron) [reply client/src/modules/assets/modals/search.modal.js]: Removed

File: client/src/modules/assets/templates/action.cell.html (hunk 9 above)
Comment (jmcameron) [reply client/src/modules/assets/templates/action.cell.html]: Done

File: server/models/migrations/next/migrate.sql (hunk 11 above)
Comment (jmcameron) [reply server/models/migrations/next/migrate.sql:91]: Done

File: server/controllers/stock/core.js (hunk 10 above)
Comment (jniles) [reply server/controllers/stock/core.js:271]: Understood.  Will make one and link it here.

**Review by jniles — CHANGES_REQUESTED** [review]
Great updates!  A few suggestions:

Could you move the indication of the asset up to the title bar?  I've also suggested an update to the text. 

image

When I click "new assignment", then cancel it, it reloads the page.  This shouldn't be necessary.  Can you check if the user has "submitted" or "canceled" and only reload the page if "submitted"?  See below:
firefox_E2c9OCvJkW

I couldn't figure out how to add a contact on the HEV production database.  I got caught in a loop where the contact group requires a contact, and the contact requires a contact type and I couldn't figure out where to put the contact type.  Could you add to the migration file:

```sql
INSERT IGNORE INTO `entity_type` (`label`, `translation_key`) VALUES    
  ('person', 'ENTITY.TYPE.PERSON'),    
  ('service', 'ENTITY.TYPE.SERVICE'),    
  ('office', 'ENTITY.TYPE.OFFICE'),    
  ('enterprise', 'ENTITY.TYPE.ENTERPRISE');   

```
to ensure we can make this work?

Can we right-align the quantity on the receipt?

image

I assigned something to myself, then unassigned it and checked the assignment history.  Here is what I saw:
image
It would be nice to make it clearer that the asset is currently unassigned.  Maybe we can add a "unassigned" in italics text after the assignment list if it is isn't assigned?

This might not be for this PR, but it would be nice to search for assets that are either (1) assigned or (2) not assigned.  So, we could add a radio button to the search modal to say "Asset Assignment" with the option:
  1. Exclude all assigned items
  2. Exclude all unassigned items

----

Overall, this looks really good.  I do think a follow up for https://github.com/IMA-WorldHealth/bhima/issues/6440 is important after this lands.  With Vanga's database, it take 5 seconds on my computer to show the asset assignment modal since it queries all lots for all time in a depot to just load the assignment modal.  This can be done after this PR lands though.

**Comment by jmcameron** [comment]
Did all the changes requested in https://github.com/IMA-WorldHealth/bhima/pull/6432#pullrequestreview-900463855.

**Review by lomamech — APPROVED** [review]
I really liked the idea of having another registry dedicated to assets, I think the PR can be merged,  I followed all the steps of the description of the PR

**Comment by lomamech** [comment]
@jniles  if you have no objection, I think we can close this PR

**Review by jniles — APPROVED** [review]
This looks really good to me.  Thanks!

bors r+

**Pull request merged**